### ISSUE TITLE
Loading partial pipeline configuration from the environment

### ISSUE THREAD

**Issue by cmanning09** [opensearch-project/data-prepper#947]
**Is your feature request related to a problem? Please describe.**
As a user of data prepper, I would like the ability to load part of my configuration from the environment. This would allow me to deploy the same pipeline configuration file into different environments and limit the number of pipeline configuration files I need to manage. My core business logic would stay the same while my environments change.

**Describe the solution you'd like**
The pipeline configuration files would support reading environment variables into the pipeline.

Example config snippet where the host url is loaded by and environment variable `OPENSEARCH_URL`:
```
sink:
   - opensearch:
     hosts:
       - ${OPENSEARCH_URL}
```

**Comment by dlvenable** [comment]
I like the idea of making pipelines more dynamic!

I'm wondering though if a properties file approach might be better though. Relying on environment variables would mean that running Data Prepper multiple times could have different results since the environment can easily change. If there was a properties file then pipeline configuration could be more dynamic and shareable, but also consistent on the same instance.

Would property files work for your use case, or is reading from environment variables necessary?

**Comment by cmanning09** [comment]
A properties file could work. Another goal of mine was trying to avoid putting some sensitive configuration data into source code like a username and password for basic auth.

**Comment by dlvenable** [comment]
Regarding sensitive data, it is generally preferable to avoid environment variables for secrets.

The Docker secrets support creates an encrypted file on the file system. It might make sense to build support for loading secrets in such a way to works with Docker secrets (but is not exclusive to it).

**Comment by cmanning09** [comment]
I agree with your points. I created a separate issue to track it. See: #949

**Comment by dlvenable** [comment]
Thanks for opening a new issue for secrets. I think there can still be value for loading from environment variables if desired.

My preference is that Data Prepper prefer using properties file over environment variables. So Data Prepper would load `${OPENSEARCH_URL}` from the properties file. It could also load `${env.OPENSEARCH_URL}` from an environment variable named `OPENSEARCH_URL`. I've seen this pattern before in other projects.

**Comment by JustinasKO** [comment]
Making config dynamic by allowing read values from env vars would make data-prepper more docker friendly. And also would give multiple options for user to solve different problems they have in regards on dynamic config. Reading some properties files is OK if you run on your local computer but probably not so good option for etc.: ECS - would need quite some hacks to inject that properties file, instead of just simple ENV vars.

> @dlvenable - Relying on environment variables would mean that running Data Prepper multiple times could have different results since the environment can easily change.

this sounds a bit funny as if you don't control your environment how can you expect to control your application?

So in conclusion having some kind of Environment substitution in config would make application more attractive for all containerisation solutions.

**Comment by icdemo** [comment]
+1 on this, the only option right now is either expose the secret or make the entire pipeline.yaml a secret.

I hope, at least the ENV_VAR will be supported.

**Comment by dlvenable** [comment]
The solution should follow the pattern from #2780 and thus take the form of `${{env:MY_ENVIRONMENT}}`, or `${{environment:MY_ENVIRONMENT}}`.

```
sink:
   - opensearch:
     hosts:
       - ${{env:OPENSEARCH_URL}}
```

**Comment by liamstar97** [comment]
@dlvenable I'm currently using data-prepper alongside Fluentd and OpenSearch in our k8s clusters at work, and we would like to use env variables for connecting to our separately namespaced OpenSearch clusters. 

eg. `opensearch.${NAMESPACE}.svc.cluster.local` or more simply, like the example you provided.

I noticed you offered guidance to another user in #2780 if they were interested in contributing. I have been considering contributing to the OpenSearch project for a while now. So if that offer still stands I'd be happy to contribute this feature to the project!

**Comment by elehcim** [comment]
This is a highly desiderable feature. Having this would simplify my work a lot.

**Comment by devops-taehoon-kim** [comment]
Any updates ?????????????????????????

**Comment by arichtman-srt** [comment]
Hi,

We achieved basically this with an initContainer running `envsubst` over the pipeline config file - you can see it here https://github.com/arichtman/data-prepper-helm-chart/blob/0571a66a9c60fd38c593a861d18ce353288a8de2/templates/deployment.yaml#L31

Hope this helps.
